wire-runtime/src/main/java/com/squareup/wire/ProtoReader.java

```diff
@@ -53,41 +54,113 @@
       "Protocol message end-group tag did not match expected tag.";
   private static final String ENCOUNTERED_A_MALFORMED_VARINT =
       "WireInput encountered a malformed varint.";
+  private static final int RECURSION_LIMIT = 64;
+
+  private final BufferedSource source;
+  private final ExtensionRegistry extensionRegistry;
+
+  /** The current position in the source, starting at 0 and increasing monotonically. */
+  private long pos = 0;
+  /** The absolute position of the end of the current message. */
+  private long currentLimit = Long.MAX_VALUE;
+  /** The current number of levels of message nesting. */
+  public int recursionDepth;
+  /** The last tag that was read. */
+  private int lastTag;
+
+  public ProtoReader(BufferedSource source) {
+    this(source, null);
+  }
+
+  // TODO ctor for extension registration
+
+  ProtoReader(BufferedSource source, ExtensionRegistry extensionRegistry) {
+    this.source = source;
+    this.extensionRegistry = extensionRegistry;
+  }
+
+  public <E> E value(TypeAdapter<E> adapter) throws IOException {
+    return adapter.read(this);
+  }
+
+  public <E> List<E> repeated(List<E> existing, TypeAdapter<E> adapter) throws IOException {
```

Comment:
Not sure if this method belongs here. Is it used frequently? I think it should be the caller's job to do the list voodoo